{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8981e9d8",
   "metadata": {},
   "source": [
    "# Generative Modeling of Isoform Expression using Diffusion Methods\n",
    "\n",
    "**Authors:**  \n",
    "Célien ABBET - s251705  \n",
    "Thibaut HEIM - s252933  \n",
    "Loïc LAISNEY - s253047  \n",
    "Yann BECKER - s253048  \n",
    "Pierre-Eduard KOLAR - s254145  \n",
    "\n",
    "**Course:** 02456 Deep Learning, DTU, Fall 2025\n",
    "\n",
    "## 0. Project Overview\n",
    "\n",
    "This notebook documents our project for the 02456 Deep Learning course.  \n",
    "The project focuses on generative modeling of isoform expression using diffusion-based methods.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f97bc9a",
   "metadata": {},
   "source": [
    "# 1. Introduction\n",
    "\n",
    "This notebook presents the full deep learning pipeline developed in our project, covering:\n",
    "\n",
    "### Training of three core models\n",
    "- Variational Autoencoder (SCimilarity VAE)\n",
    "- Diffusion model (scDiffusion)\n",
    "- Cell-type classifier\n",
    "\n",
    "### Application of the trained models for\n",
    "- Bulk isoform inference (guided/non guided)\n",
    "- Single-cell isoform inference (guided/non guided)\n",
    "\n",
    "### A set of evaluation metrics\n",
    "- MMD, KL divergence, Wasserstein distances\n",
    "- Random Forest distinguishability test. We measure the AUC (Area Under Curve)\n",
    "- UMAP and ROC visualization\n",
    "\n",
    "---\n",
    "\n",
    "The notebook does not execute heavy computations. Instead, it documents the methods and illustrates outputs using figures generated beforehand, during and after model development.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5aaaef94",
   "metadata": {},
   "source": [
    "# 2. Datasets\n",
    "\n",
    "## 2.1 Bulk Dataset\n",
    "- Transcriptome isoform quantification  \n",
    "- Input dimensionality: 162,009 features\n",
    "- Leiden clustering used to derive reference cell-type labels  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "**Bulk transcripts dataset Overview**\n",
    "```json\n",
    "============================================================\n",
    "Dataset: bulk_transcripts\n",
    "============================================================\n",
    "\n",
    "Shape: 19882 cells × 162009 features\n",
    "  - n_obs (cells): 19882\n",
    "  - n_vars (genes/transcripts): 162009\n",
    "\n",
    "Data Matrix (X):\n",
    "  - Type: <class 'scipy.sparse._csr.csr_matrix'>\n",
    "  - Dtype: uint32\n",
    "  - Memory: 6.85 GB\n",
    "  - Sparse: True\n",
    "\n",
    "Cell metadata (obs): ['geo_accession', 'series_id', 'characteristics_ch1', 'extract_protocol_ch1', 'source_name_ch1', 'title', 'contact_city', 'contact_country', 'contact_institute', 'instrument_model', 'library_source', 'organism_ch1', 'platform_id', 'singlecellprobability', 'submission_date', 'taxid_ch1', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'mt_outlier', 'n_counts', 'n_genes', 'leiden']\n",
    "Gene/Transcript metadata (var): ['gene_name', 'gene_id']\n",
    "Layers: None\n",
    "Obs metadata (obsm): None\n",
    "Var metadata (varm): None\n",
    "Unstructured annotations (uns): ['gene_n_transcripts', 'gene_to_transcripts', 'multi_isoform_genes', 'single_isoform_genes', 'transcript_id_to_index', 'transcript_mapping']\n",
    "\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77c634c2",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "## 2.2 Single-Cell Dataset\n",
    "- Single-cell isoform expression matrix  \n",
    "- Input dimensionality: 179,610 features  \n",
    "- Leiden clustering used to derive reference cell-type labels  \n",
    "\n",
    "**Single-cell transcripts dataset Overview**\n",
    "\n",
    "```json\n",
    "============================================================\n",
    "Dataset: sc_transcripts\n",
    "============================================================\n",
    "\n",
    "Shape: 183880 cells/genes × 179610 features\n",
    "  - n_obs (cells): 183880\n",
    "  - n_vars (genes/transcripts): 179610\n",
    "\n",
    "Data Matrix (X):\n",
    "  - Type: <class 'scipy.sparse._csr.csr_matrix'>\n",
    "  - Dtype: uint32\n",
    "  - Memory: 9.72 GB\n",
    "  - Sparse: True\n",
    "\n",
    "Cell metadata (obs): ['geo_accession', 'series_id', 'characteristics_ch1', 'extract_protocol_ch1', 'source_name_ch1', 'title', 'contact_city', 'contact_country', 'contact_institute', 'instrument_model', 'library_source', 'organism_ch1', 'platform_id', 'singlecellprobability', 'submission_date', 'taxid_ch1', 'n_genes_by_counts', 'total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'mt_outlier', 'n_counts', 'n_genes', 'leiden']\n",
    "Gene/Transcript metadata (var): ['gene_name', 'gene_id']\n",
    "Layers: None\n",
    "Obs metadata (obsm): None\n",
    "Var metadata (varm): None\n",
    "Unstructured annotations (uns): ['gene_n_transcripts', 'gene_to_transcripts', 'multi_isoform_genes', 'single_isoform_genes', 'transcript_id_to_index', 'transcript_mapping']\n",
    "```\n",
    "<br><br>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70de90b5",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "Most metadata will not be used in this project. Although, the leiden clusters (``` 'leiden' ```) will be valuable for conditioning the generation of isoform expressions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00108ee2",
   "metadata": {},
   "source": [
    "\n",
    "## 2.3 Preprocessing\n",
    "- Normalization\n",
    "- Log1p transformation\n",
    "\n",
    "\n",
    "```python\n",
    "# Loading the data\n",
    "    adata = ad.read_h5ad(\"/work3/s193518/scIsoPred/data/bulk_processed_transcripts.h5ad\")\n",
    "# Library size normalization\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "# Log1p transformation\n",
    "    sc.pp.log1p(adata)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd01defd",
   "metadata": {},
   "source": [
    "# 3. Variational Autoencoder (VAE)\n",
    "\n",
    "## 3.1 VAE Architecture\n",
    "The VAE architecture is based on SCimilarity.Their Zenodo records let us download some useful code.\n",
    "\n",
    "We used their annotation model (V1) as a basis for our architecture (`url = \"https://zenodo.org/record/8286452/files/annotation_model_v1.tar.gz?download=1\"`)\n",
    "\n",
    "Their weights can be downloaded with `src/utils/download_VAE.py` for finetuning.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c60a2734",
   "metadata": {},
   "source": [
    "\n",
    "- The obtained architecture can be seen in `src/VAE/VAE_model.py`\n",
    "\n",
    "```python\n",
    "class Encoder(nn.Module):\n",
    "    ...\n",
    "\n",
    "class Decoder(nn.Module):\n",
    "    ...\n",
    "\n",
    "class VAE(torch.nn.Module):\n",
    "    ...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a24d270",
   "metadata": {},
   "source": [
    "## 3.2 Training Procedure\n",
    "- Training script: `src/VAE/VAE_train.py`\n",
    "\n",
    "**Command example:**\n",
    "```bash\n",
    "python VAE_train.py \\\n",
    "    --data_dir '/work3/s193518/scIsoPred/data/sc_processed_transcripts.h5ad' \\\n",
    "    --num_genes 179610 \\\n",
    "    --max_steps 4000 \\\n",
    "    --max_minutes 3000 \\\n",
    "    --checkpoint_freq 1000 \\\n",
    "    --batch_size 128 \\\n",
    "    --state_dict \"./annotation_model_v1\" \\\n",
    "    --save_dir \"./output/ae_checkpoint/vae_sc_transcript/\"\n",
    "```\n",
    "**Note**: The `state_dict` argument specifies the path to the initial weights (when finetuning).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abd39833",
   "metadata": {},
   "source": [
    "# 4. Diffusion Model (scDiffusion)\n",
    "\n",
    "\n",
    "## 4.1 Model Architecture\n",
    "\n",
    "We once again use an existing architecture. The scDiffusion model can be found on github: `https://github.com/EperLuo/scDiffusion/tree/main`\n",
    "\n",
    "- Source file: `src/classifier/cell_model.py`\n",
    "\n",
    "```python\n",
    "class Cell_Unet(nn.Module):\n",
    "    ...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8d52c25",
   "metadata": {},
   "source": [
    "## 4.2 Training Procedure\n",
    "- Training script: `src/diffusion/diffusion_train.py`\n",
    "\n",
    "We trained three variants:  \n",
    "- Bulk-only training  \n",
    "- Bulk pretraining followed by single-cell finetuning  \n",
    "- Single-cell training from scratch  \n",
    "\n",
    "**Command example:**  \n",
    "```bash\n",
    "python diffusion_train.py \\\n",
    "    --data_dir \"/work3/s193518/scIsoPred/data/bulk_processed_transcripts.h5ad\" \\\n",
    "    --lr_anneal_steps 200000 \\\n",
    "    --batch_size 128 \\\n",
    "    --log_interval 10 \\\n",
    "    --save_interval 1000 \\\n",
    "    --save_interval 100000 \\\n",
    "    --vae_path /zhome/5b/d/223428/DTU_DL_PROJECT_DIFFUSION/src/VAE/output/ae_checkpoint/vae_sc_transcript/model_seed=0_step=3999.pt' \\\n",
    "    --latent_dim 128 \\\n",
    "    --model_path 'output/classifier_checkpoint/classifier_sc_processed_transcripts' \\\n",
    "    --start_guide_time 500 \\\n",
    "    --num_class 44\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4603a97a",
   "metadata": {},
   "source": [
    "## 4.3 Inference\n",
    "\n",
    "- Inference script: `src/diffusion/cell_sample.py`\n",
    "\n",
    "\n",
    "**Command example**\n",
    "```bash\n",
    "python cell_sample.py \\\n",
    "    --num_samples 12000 \\\n",
    "    --batch_size 3000 \\\n",
    "    --model_path \"model.pt\" \\\n",
    "    --sample_dir \"samples\"\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0fd064d",
   "metadata": {},
   "source": [
    "## 4.4 Visualization (UMAP, ROC)\n",
    "\n",
    "Click for full size\n",
    "\n",
    "\n",
    "**Bulk-only**\n",
    "\n",
    "<a href=\"umap/UMAP_Global_bulk_non_guided.png\" target=\"_blank\">\n",
    "  <img src=\"umap/UMAP_Global_bulk_non_guided.png\" alt=\"Isoform expression\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/random_forest_plot_Global_bulk_ng.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/random_forest_plot_Global_bulk_ng.png\" alt=\"Isoform expression\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "**Single-cell only**\n",
    "\n",
    "<a href=\"umap/UMAP_Global_sc_non_guided.png\" target=\"_blank\">\n",
    "  <img src=\"umap/UMAP_Global_sc_non_guided.png\" alt=\"Isoform expression\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/random_forest_plot_sc_non_guided.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/random_forest_plot_sc_non_guided.png\" alt=\"Isoform expression\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "\n",
    "**Single-cell transfer learning**\n",
    "\n",
    "<a href=\"umap/UMAP_Globa_sc_transfer_non_guided.png\" target=\"_blank\">\n",
    "  <img src=\"umap/UMAP_Globa_sc_transfer_non_guided.png\" alt=\"Isoform expression\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/random_forest_plot_transfer_non_guided.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/random_forest_plot_transfer_non_guided.png\" alt=\"Isoform expression\" height=\"250\">\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7307dc6",
   "metadata": {},
   "source": [
    "\n",
    "**Single-cell unique class**\n",
    "\n",
    "<a href=\"umap/sc_unique_class/UMAP_Unique_Class_0.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_unique_class/UMAP_Unique_Class_0.png\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/sc_unique_class/random_forest_plot_Unique_Class_0_sc_False_False.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/sc_unique_class/random_forest_plot_Unique_Class_0_sc_False_False.png\" height=\"250\">\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cd0e9b2",
   "metadata": {},
   "source": [
    "# 5. Leiden-Cluster Classifier\n",
    "\n",
    "The classifier is also an adaptation from scDiffusion.\n",
    "\n",
    "## 5.1 Architecture\n",
    "- Source file: `src/classifier/cell_model.py`\n",
    "\n",
    "```python\n",
    "class Cell_classifier(nn.Module):\n",
    "    ...\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b009c0ae",
   "metadata": {},
   "source": [
    "## 5.2 Training Procedure\n",
    "- Training script: `src/classifier/classifier_train.py`\n",
    "\n",
    "**Command example:**\n",
    "```bash\n",
    "python classifier_train.py \\\n",
    "    --data_dir \"/work3/s193518/scIsoPred/data/sc_processed_transcripts.h5ad\" \\\n",
    "    --iterations 200000 \\\n",
    "    --batch_size 128 \\\n",
    "    --log_interval 20 \\\n",
    "    --eval_interval 20 \\\n",
    "    --save_interval 100000 \\\n",
    "    --vae_path /zhome/5b/d/223428/DTU_DL_PROJECT_DIFFUSION/src/VAE/output/ae_checkpoint/vae_sc_transcript/model_seed=0_step=3999.pt' \\\n",
    "    --latent_dim 128 \\\n",
    "    --model_path 'classifier_sc_processed_transcripts' \\\n",
    "    --start_guide_time 500 \\\n",
    "    --num_class 44\n",
    "```\n",
    "**Note**: In the diffusion process, we use 1000 noise steps.  \n",
    "The argument `--start_guide_time 500` in the classifier corresponds to the point in the reverse diffusion trajectory at which classifier guidance is activated. Since early timesteps (close to 1000) are dominated by pure noise, the classifier signal would be unreliable there. By starting guidance at step 500, we make sure that the generated samples have denoised sufficiently for the classifier to provide stable and meaningful gradients.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0352480f",
   "metadata": {},
   "source": [
    "## 5.3 Inference\n",
    "\n",
    "- Inference script: `src/diffusion/classifier_sample.py`\n",
    "\n",
    "\n",
    "**Command example**\n",
    "```bash\n",
    "python classifier_sample.py \\\n",
    "    --num_samples 9000 \\\n",
    "    --batch_size 3000 \\\n",
    "    --model_path \"output/diffusion/diff_model.pt\" \\\n",
    "    --classifier_path \"output/classifier/class_model.pt\" \\\n",
    "    --sample_dir \"output/simulated_samples/single_cell\" \\\n",
    "    --start_guide_steps 500 \\\n",
    "    --classifier_scale 2\n",
    "```\n",
    "**Note**: The argument `--classifier_scale` denotes the weight of the guidance in the generation of features (the variable *s* in the equation (3) of the report). Results will show that a value of 2 might not be the best."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a332fbc1",
   "metadata": {},
   "source": [
    "## 5.4 Visualization (UMAP, ROC)\n",
    "\n",
    "Click for full size\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f5a76e",
   "metadata": {},
   "source": [
    "**Bulk-only**\n",
    "\n",
    "\n",
    "<a href=\"umap/bulk_guided/UMAP_Cluster_0.png\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/UMAP_Cluster_0.png\" alt=\"UMAP Cluster 0\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/bulk_guided/UMAP_Cluster_1.png\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/UMAP_Cluster_1.png\" alt=\"UMAP Cluster 1\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/bulk_guided/UMAP_Cluster_2.png\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/UMAP_Cluster_2.png\" alt=\"UMAP Cluster 2\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/bulk_guided/UMAP_Cluster_3.png\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/UMAP_Cluster_3.png\" alt=\"UMAP Cluster 3\" width=\"250\">\n",
    "</a>\n",
    "\n",
    "<br>\n",
    "\n",
    "\n",
    "<a href=\"umap/bulk_guided/UMAP_Combined_Clusters_0_1_2_3_4_new.PNG\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/UMAP_Combined_Clusters_0_1_2_3_4_new.PNG\" alt=\"Combined Clusters\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/bulk_guided/random_forest_plot_Guided_Global_bulk_True_False.png\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/random_forest_plot_Guided_Global_bulk_True_False.png\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/bulk_guided/random_forest_plot_Cluster_0_bulk_True_False.png\" target=\"_blank\">\n",
    "  <img src=\"umap/bulk_guided/random_forest_plot_Cluster_0_bulk_True_False.png\" height=\"250\">\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d57a00",
   "metadata": {},
   "source": [
    "**Single-cell only**\n",
    "\n",
    "<a href=\"umap/sc_guided/UMAP_Cluster_0.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_guided/UMAP_Cluster_0.png\" alt=\"UMAP Cluster 0\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/sc_guided/UMAP_Cluster_6.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_guided/UMAP_Cluster_6.png\" alt=\"UMAP Cluster 6\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/sc_guided/UMAP_Cluster_7.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_guided/UMAP_Cluster_7.png\" alt=\"UMAP Cluster 7\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/sc_guided/UMAP_Cluster_12.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_guided/UMAP_Cluster_12.png\" alt=\"UMAP Cluster 12\" width=\"250\">\n",
    "</a>\n",
    "<a href=\"umap/sc_guided/UMAP_Cluster_18.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_guided/UMAP_Cluster_18.png\" alt=\"UMAP Cluster 18\" width=\"250\">\n",
    "</a>\n",
    "\n",
    "<br>\n",
    "\n",
    "<a href=\"umap/sc_guided/UMAP_Combined_Clusters_0_1_2_3_4_5_6_7_8.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_guided/UMAP_Combined_Clusters_0_1_2_3_4_5_6_7_8.png\" alt=\"Combined Clusters\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/sc_guided/random_forest_plot_sc_Guided_Global.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/sc_guided/random_forest_plot_sc_Guided_Global.png\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/sc_guided/random_forest_sc_guided_plot_Cluster_0.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/sc_guided/random_forest_sc_guided_plot_Cluster_0.png\" height=\"250\">\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c4aa720",
   "metadata": {},
   "source": [
    "\n",
    "**Single-cell transfer learning**\n",
    "\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_0.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_0.png\" alt=\"Cluster 0\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_1.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_1.png\" alt=\"Cluster 1\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_2.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_2.png\" alt=\"Cluster 2\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_4.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_4.png\" alt=\"Cluster 4\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_6.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_6.png\" alt=\"Cluster 6\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_7.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_7.png\" alt=\"Cluster 7\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_8.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_8.png\" alt=\"Cluster 8\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_10.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_10.png\" alt=\"Cluster 10\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_12.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_12.png\" alt=\"Cluster 12\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_13.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_13.png\" alt=\"Cluster 13\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_14.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_14.png\" alt=\"Cluster 14\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_16.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_16.png\" alt=\"Cluster 16\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_17.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_17.png\" alt=\"Cluster 17\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_18.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_18.png\" alt=\"Cluster 18\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_19.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_19.png\" alt=\"Cluster 19\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_20.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_20.png\" alt=\"Cluster 20\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_21.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_21.png\" alt=\"Cluster 21\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_22.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_22.png\" alt=\"Cluster 22\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_23.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_23.png\" alt=\"Cluster 23\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_24.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_24.png\" alt=\"Cluster 24\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_26.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_26.png\" alt=\"Cluster 26\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_27.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_27.png\" alt=\"Cluster 27\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_28.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_28.png\" alt=\"Cluster 28\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_29.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_29.png\" alt=\"Cluster 29\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_30.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_30.png\" alt=\"Cluster 30\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_31.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_31.png\" alt=\"Cluster 31\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/ROC_rd_forests/sc_unique_class/chien_mechant.jpeg\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_32.png\" alt=\"Cluster 32\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_33.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_33.png\" alt=\"Cluster 33\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_34.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_34.png\" alt=\"Cluster 34\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_35.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_35.png\" alt=\"Cluster 35\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_36.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_36.png\" alt=\"Cluster 36\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_37.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_37.png\" alt=\"Cluster 37\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_38.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_38.png\" alt=\"Cluster 38\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_39.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_39.png\" alt=\"Cluster 39\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_42.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_42.png\" alt=\"Cluster 42\" width=\"150\">\n",
    "</a>\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Cluster_43.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Cluster_43.png\" alt=\"Cluster 43\" width=\"150\">\n",
    "</a>\n",
    "\n",
    "<br>\n",
    "\n",
    "<a href=\"umap/sc_transfer_guided/UMAP_Combined_Clusters_0_1_2_3_4_5_7_8.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_transfer_guided/UMAP_Combined_Clusters_0_1_2_3_4_5_7_8.png\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/sc_transfer_guided/random_forest_plot_Guided_Global_sc_True_True.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/sc_transfer_guided/random_forest_plot_Guided_Global_sc_True_True.png\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/sc_transfer_guided/random_forest_plot_Cluster_0_sc_True_True.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/sc_transfer_guided/random_forest_plot_Cluster_0_sc_True_True.png\" height=\"250\">\n",
    "</a>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bab6e46",
   "metadata": {},
   "source": [
    "\n",
    "**Single-cell unique class**\n",
    "\n",
    "<a href=\"umap/sc_unique_class/UMAP_Unique_Class_0.png\" target=\"_blank\">\n",
    "  <img src=\"umap/sc_unique_class/UMAP_Unique_Class_0.png\" height=\"250\">\n",
    "</a>\n",
    "\n",
    "<a href=\"umap/ROC_rd_forests/sc_unique_class/random_forest_plot_Unique_Class_0_sc_False_False.png\" target=\"_blank\">\n",
    "  <img src=\"umap/ROC_rd_forests/sc_unique_class/random_forest_plot_Unique_Class_0_sc_False_False.png\" height=\"250\">\n",
    "</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "940613d6",
   "metadata": {},
   "source": [
    "# 6. Evaluation Metrics\n",
    "\n",
    "We use the metrics listed in section 3.3.2 of the report.\n",
    "\n",
    "- File to run the metrics: `src/metrics/run_metrics_generator.py`\n",
    "\n",
    "**Command example**\n",
    "```bash\n",
    "python run_metrics_generator.py \\\n",
    "    --mode 'sc' \\\n",
    "    --guided True \\\n",
    "    --transfer True \\\n",
    "    --num_samples 1000 \\\n",
    "    --sample_dir \"output/simulated_samples/single_cell\" \\\n",
    "    --start_guide_steps 500 \\\n",
    "    --rf True \\\n",
    "    --per_cluster True\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5ca5f0f",
   "metadata": {},
   "source": [
    "## 6.1 Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62757595",
   "metadata": {},
   "source": [
    "| **Metrics**     | **Bulk NG** | **Bulk G** | **Sc NG** | **Sc G** | **TL NG** | **TL G** |\n",
    "|-----------------|------------:|-----------:|----------:|---------:|----------:|---------:|\n",
    "| **MMD**         | 4.0e-4      | 5.8e-4     | 4.0e-4    | 5.2e-4   | 4.0e-4    | 4.0e-4   |            \n",
    "| **Wasserstein** | 1.2         | 1.1        | 1.6       | 2.6      | 2.1       | 2.0      |           \n",
    "| **KL**          | 1.7         | 2.6        | 3.3       | 13       | 8.2       | 7.7      |        \n",
    "| **AUC**         | 0.94        | 0.96       | 0.98      | 0.99     | 1.0       | 1.0      |\n",
    "\n",
    "**Note:** All those metrics are measured when generating global features. The following comparison focuses on generated features when conditioning on the Leiden cluster 0.\n",
    "\n",
    "| **Metrics**     | **Sc G (Cluster 0)** | **1 Class NG (Cluster 0)** |\n",
    "|-----------------|---------------------:|---------------------------:|\n",
    "| **MMD**         | 2.0e-3               | 1.4e-3             \n",
    "| **Wasserstein** | 1.56                 | 1.7             \n",
    "| **KL**          | 11                   | 14             \n",
    "| **AUC**         | 1.0                  | 0.93"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5fa06e",
   "metadata": {},
   "source": [
    "# 7. Conclusion\n",
    "\n",
    "Overall, the project demonstrates that diffusion-based generative modeling can capture complex isoform expression structures, but several limitations remain. First, transfer learning did not yield the expected improvements, likely because bulk and single-cell distributions are too different for effective transfer. Second, the classifier-guidance mechanism appears suboptimal, the chosen guidance parameter may not be well aligned with the diffusion dynamics. Further discussion on the results and the overall project can be found in Section 4 of the report."
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
